crates/router/src/db/customers.rs

```diff
@@ -63,9 +70,298 @@ where
+        let database_call = || async {
+            storage::Customer::find_optional_by_customer_id_merchant_id(
+                &conn,
+                customer_id,
+                merchant_id,
+            )
+            .await
+            .map_err(Into::into)
+            .into_report()
+        };
+
+        let maybe_customer = match storage_scheme {
+            MerchantStorageScheme::PostgresOnly => database_call().await,
+            MerchantStorageScheme::RedisKv => {
+                let key = format!("mid_{}_cust_{}", merchant_id, customer_id);
+                let field = format!("cust_{}", customer_id);
+                Box::pin(db_utils::try_redis_get_else_try_database_get( // check for ValueNotFound
+                    async {
+                        kv_wrapper(self, KvOperation::<diesel_models::Customer>::HGet(&field),key)
+                            .await?
+                            .try_into_hget()
+                            .map(|ok_res| Some(ok_res))
+                    },
+                    database_call,
+                ))
+                .await
+            }
+        }?;
+
+        let maybe_result = maybe_customer.async_map(|c| async{
+            c.convert(key_store.key.get_inner())
+                .await
+                .change_context(errors::StorageError::DecryptionError)
+        }).await.transpose()?;
+
+        maybe_result.map_or(Ok(None), |customer:domain::Customer| {
+            match customer.name {
+                Some(ref name) if name.peek() == REDACTED => {
+                    Err(errors::StorageError::CustomerRedacted)?
+                }
+                _ => Ok(Some(customer)),
+            }
+        })
+    }
+
+    #[instrument(skip_all)]
+    async fn update_customer_by_customer_id_merchant_id(
+        &self,
+        customer_id: String,
+        merchant_id: String,
+        customer : domain::Customer,
+        customer_update: storage::CustomerUpdate,
+        key_store: &domain::MerchantKeyStore,
+        storage_scheme: MerchantStorageScheme,
+    ) -> CustomResult<domain::Customer, errors::StorageError> {
+        let conn = connection::pg_connection_write(self).await?;
+        let customer = Conversion::convert(customer)
+                .await
+                .change_context(errors::StorageError::EncryptionError)?;
+        let database_call = || async {
+            storage::Customer::update_by_customer_id_merchant_id(
+                &conn,
+                customer_id.clone(),
+                merchant_id.clone(),
+                customer_update.clone().into()
+            )
+            .await
+            .map_err(Into::into)
+            .into_report()
+        };
+
+        let updated_object = match storage_scheme {
+            MerchantStorageScheme::PostgresOnly => database_call().await,
+            MerchantStorageScheme::RedisKv => {
+                let key = format!("mid_{}_cust_{}", merchant_id, customer_id);
```

Comment:
Should the key be merchant_id_customer_id? I dont think this serves the purpose of sharding all the data with single key. Should we think of something else

cc: @jarnura

Reply:
Because I think KV is already there for Payment related customer data, which has the shard key as merchant_id_payment_id

Reply:
i was thinking mandate data can also come under mid_cust_id

Reply:
But for address it's mid_pid because only payment related address is going to KV

Reply:
how does that work, would address be changed or created only in payments flow

Reply:
Yes the addresses which are going through KV flow are only inserted during payments